File: Data/1_FeatureEngineering.r
library(SmartEDA)
library(arrow)
library(tibble)
library(ggplot2)
library(ggthemes)
library(ggbreak)
library(svglite)
library(dplyr)
library(tidyr)
library(texreg)
library(xtable)
library(beepr)
library(readxl)
library(MASS)
library(purrr)
library(arrow)
library(mice)

rm(list=ls())

# Source GetData.
source('0_GetData.r')

# Source theme_elcon
invisible(source('theme_elcon.R'))

dir <- "C:/Users/tobr/OneDrive - NRGi A S/Projekter/ProjectBasedInternship/Data"
setwd(dir)

dfData <- arrow::read_parquet("./dfData.parquet")

# Order by date
dfData <- dfData %>% arrange(date)

# Summary of Data
eda_1 <- xtable(ExpData(data=dfData,type=1),
            caption= "Summary of Dataset",
            align=c("l","l","c"), include.rownames=F)
print(eda_1,file="./Results/Tables/eda_1.tex",append=F,
      caption.placement="top")
eda_1

# Summary of Variables
eda_2 <- xtable(ExpData(data=dfData,type=2,fun = c('mean'))[,-1],
              caption= "Summary of Variables", include.rownames=F)
print(eda_2,file="./Results/Tables/eda_2.tex",append=F,
      caption.placement="top",floating = T, floating.environment = "sidewaystable")
eda_2

# Summary of Categorical Variables
eda_3 <- xtable(ExpCTable(dfData,Target="department",margin=1,clim=10,nlim=3,round=2,bin=NULL,per=T),
                caption= "Summary of Categorical Variables by Deparment", include.rownames=F)
print(eda_3,file="./Results/Tables/eda_3.tex",append=F,
      caption.placement="top",floating = T, floating.environment = "sidewaystable")
eda_3

### Explore numeric variables ###
# Plot correlation matrix
# Transform to cross-sectional data using group by job_no, calculate cumulative numbers, and select latest date
colNum <- names(dfData)[sapply(dfData, is.numeric)]
colNum <- colNum[!grepl("_share",colNum)]
colNum <- colNum[!grepl("_rate",colNum)]
colNum <- colNum[!grepl("_ratio",colNum)]
colNum <- colNum[!grepl("_margin",colNum)]
#colNum <- colNum[!grepl("_estimate",colNum)]
colNum <- colNum[!grepl("_cumsum",colNum)]
colCumSum <- colNum[!grepl("budget_",colNum)]


# For each variable in colCumSum create a new variable with cumulative sum by job_no. Name of new variable is the same as the old variable with "_cumsum" added
dfData <- dfData %>%
  group_by(job_no) %>%
  arrange(date) %>%
  mutate(across(all_of(colCumSum), cumsum, .names = "{.col}_cumsum")) %>%
  ungroup()

# Calculate days until end of job
dfData <- dfData %>%
  mutate(days_until_end = end_date - date)

dfData$days_until_end[dfData$days_until_end < 0] <- 0
dfData$days_until_end <- as.numeric(dfData$days_until_end)

##### Feature engineering #####
# Change in sales_estimate_contribution from previous observation. Group by job_no and order by date
dfData <- dfData %>%
  group_by(job_no) %>%
  arrange(date) %>%
  mutate(sales_estimate_contribution_change = sales_estimate_contribution - lag(sales_estimate_contribution),
         production_estimate_contribution_change = production_estimate_contribution - lag(production_estimate_contribution),
         final_estimate_contribution_change = final_estimate_contribution - lag(final_estimate_contribution))

dfData <- dfData %>%
  group_by(job_no) %>%
  arrange(date) %>%
  mutate(sales_estimate_contribution_change = ifelse(is.na(sales_estimate_contribution_change), 0, sales_estimate_contribution_change),
         production_estimate_contribution_change = ifelse(is.na(production_estimate_contribution_change), 0, production_estimate_contribution_change),
         final_estimate_contribution_change = ifelse(is.na(final_estimate_contribution_change), 0, final_estimate_contribution_change))

# Calculate revenue according to s-curve
dfData <- dfData %>%
  group_by(job_no) %>%
  arrange(date) %>%
  mutate(days_since_start = date - min(date),
         total_days = max(end_date) - min(date), # make total_days a difftime object
         progress = as.numeric(days_since_start) / as.numeric(total_days)) %>%
         ungroup() %>% # S-curve according to (1/(1+exp(-kx))^a.
         mutate(
         scurve = (1/(1+exp(-6*(progress-0.5))))^2,
         revenue_scurve = scurve * budget_revenue,
         costs_scurve = scurve * budget_costs,
         revenue_scurve_diff = revenue_scurve - revenue_cumsum,
         costs_scurve_diff = costs_scurve - costs_cumsum,
         contribution_scurve = scurve * (budget_revenue - budget_costs),
         contribution_scurve_diff = contribution_scurve - contribution_cumsum
         )

# Read xlxs file with cvr, equity ratio and quick ratio
dfCvr <- read_excel("./.AUX/Cvr.xlsx")
dfCvr <- dfCvr %>%
  na.omit()
# Drop column 2
dfCvr <- dfCvr[,-2]
# Rename columns
colnames(dfCvr) <- c("cvr","customer_equity_ratio","customer_quick_ratio")
# Set cvr as character
dfCvr$cvr <- as.character(dfCvr$cvr)

# Replace . with , in cvr
dfCvr$customer_equity_ratio <- as.numeric(gsub("\\,","\\.",dfCvr$customer_equity_ratio))
dfCvr$customer_quick_ratio <- as.numeric(gsub("\\,","\\.",dfCvr$customer_quick_ratio))

# Divide by 100
dfCvr$customer_equity_ratio <- dfCvr$customer_equity_ratio/100
dfCvr$customer_quick_ratio <- dfCvr$customer_quick_ratio/100

# Join dfCvr on dfData
dfData <- left_join(dfData,dfCvr,by="cvr")

# Impute missing values in customer_equity_ratio and customer_quick_ratio with mean
dfData$customer_equity_ratio[is.na(dfData$customer_equity_ratio)] <- mean(dfData$customer_equity_ratio, na.rm = T)
dfData$customer_quick_ratio[is.na(dfData$customer_quick_ratio)] <- mean(dfData$customer_quick_ratio, na.rm = T)

dfOutstanding <- read_excel("./.AUX/Debitorer.xlsx")
dfOutstanding <- dfOutstanding %>%
  na.omit()
# Drop column 3
dfOutstanding <- dfOutstanding[,-3]
# Rename columns
colnames(dfOutstanding) <- c("date","department","outstanding","cvr")
# Set cvr as character
dfOutstanding$cvr <- as.character(dfOutstanding$cvr)
dfOutstanding$department <- as.factor(dfOutstanding$department)

dfData <- left_join(dfData,dfOutstanding,by=c("cvr"="cvr","department"="department","date"="date"))

dfData$outstanding[is.na(dfData$outstanding)] <- 0

# Replace NaN & Inf with NA
dfData <- replace(dfData, is.infinite(as.matrix(dfData)), NA)
dfData <- replace(dfData, is.nan(as.matrix(dfData)), NA)

# Replace NaN with NA
dfData <- replace(dfData, is.nan(as.matrix(dfData)), NA)
# Replace Inf with NA
dfData <- replace(dfData, is.infinite(as.matrix(dfData)), NA)

# Get the last observation of each job_no and determine if total contribution is positive or negative
dfDataContrib <- dfData %>%
  group_by(job_no) %>%
  mutate_at(colNum, cumsum) %>%
  filter(date == max(date))

# If contribution_cumsum is positive, set contribution_positive to 1, else set to 0
dfDataContrib$profitable <- ifelse(dfDataContrib$contribution_cumsum > 0, 1, 0)

# Omit all columns except job_no and profitable
dfDataContrib <- dfDataContrib[,c("job_no","profitable")]

# Join dfDataContrib on dfData
dfData <- left_join(dfData,dfDataContrib,by="job_no")

# Define risk as the log of the ratio between the realized and the S-curve. Omit NaN
dfData <- dfData %>%
  group_by(job_no) %>%
  arrange(date) %>%
  mutate(
    efficiency_risk = (costs_scurve_diff * costs_of_labor_share -(billable_rate_dep -0.90) + illness_rate_dep
                               + internal_rate_dep + earned_time_off_rate + allowance_rate)^(2),
    overrun_risk = (costs_scurve_diff * costs_of_materials_share + costs_of_materials_share * billable_rate_dep
                                  + costs_of_materials_share)^(2)
  ) %>%
  do({
    if (any(is.na(.$contribution_scurve_diff)) || any(is.na(.$contribution_cumsum))) {
      data.frame(., risk = NA_real_)
    } else {
      model <- lm(contribution_scurve_diff ~ revenue_scurve_diff + costs_scurve_diff + 1 +
                                             billable_rate_dep + illness_rate_dep + internal_rate_dep
      , data = .)
      data.frame(., risk = exp(model$residuals))
    }
  })


library(ExPanDaR)

prepare_scatter_plot(dfData, x="risk", y="contribution", color="department", loess = 1)

cor(dfData$risk,dfData$contribution,use="complete.obs")

File: Data/Scripts/1_FeatureEngineering.py
import os
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.feature_extraction.text import CountVectorizer
from nltk.corpus import stopwords
from nltk.stem.snowball import DanishStemmer
import re
from plot_config import *

# Set working directory
sDir = "C:/Users/tobr/OneDrive - NRGi A S/Projekter/ProjectBasedInternship/Data"
os.chdir(sDir)

# Read data
dfData = pd.read_parquet(f"{sDir}/dfData.parquet")

# Order by date
dfData.sort_values('date', inplace=True)

### Explore numeric variables ###
# Identify numeric columns based on conditions
colNum = [col for col in dfData.select_dtypes(include=[np.number]).columns if
          not any(sub in col for sub in ["_share", "_rate", "_ratio", "_margin", "_cumsum"])]
colCumSum = [col for col in colNum if "budget_" not in col]

# Calculate cumulative sum for each variable in colCumSum grouped by 'job_no'
for col in colCumSum:
    dfData[f'{col}_cumsum'] = dfData.groupby('job_no')[col].cumsum()

# Calculate days until end of job
dfData['days_until_end'] = (dfData['end_date'] - dfData['date']).dt.days
dfData.loc[dfData['days_until_end'] < 0, 'days_until_end'] = 0

##### Feature engineering #####
# Calculate change in various estimates
for estimate_type in ['sales', 'production', 'final']:
    dfData[f'{estimate_type}_estimate_contribution_change'] = dfData.groupby('job_no')[
        f'{estimate_type}_estimate_contribution'].diff().fillna(0)

# S-curve calculations
dfData['days_since_start'] = (dfData['date'] - dfData.groupby('job_no')['date'].transform('min')).dt.days
dfData['total_days'] = (
        dfData.groupby('job_no')['end_date'].transform('max') - dfData.groupby('job_no')['date'].transform(
    'min')).dt.days
dfData['progress'] = dfData['days_since_start'] / dfData['total_days']

k = 6  # Coefficient for S-curve
a = 2  # Exponent for S-curve
dfData['scurve'] = (1 / (1 + np.exp(-k * (dfData['progress'] - 0.5)))) ** a
dfData['revenue_scurve'] = dfData['scurve'] * dfData['budget_revenue']
dfData['costs_scurve'] = dfData['scurve'] * dfData['budget_costs']
dfData['revenue_scurve_diff'] = dfData['revenue_scurve'] - dfData['revenue_cumsum']
dfData['costs_scurve_diff'] = dfData['costs_scurve'] - dfData['costs_cumsum']
dfData['contribution_scurve'] = dfData['scurve'] * (dfData['budget_revenue'] - dfData['budget_costs'])
dfData['contribution_scurve_diff'] = dfData['contribution_scurve'] - dfData['contribution_cumsum']

# Plot distribution of budget_costs, sales_estimate_costs, production_estimate_costs and final_estimate_costs
plt.figure(figsize=(10, 5))
sns.kdeplot(data=dfData, x='budget_costs', label='budget costs')
sns.kdeplot(data=dfData, x='sales_estimate_costs', label='sales estimate costs')
sns.kdeplot(data=dfData, x='production_estimate_costs', label='production estimate costs')
sns.kdeplot(data=dfData, x='final_estimate_costs', label='final estimate costs')
plt.xlabel("Costs (mDKK)")
plt.ylabel("Density")
plt.xlim(dfData['budget_costs'].quantile(0.00000000001), 10)
plt.legend(loc='upper center', bbox_to_anchor=(0.5, -0.1), ncol=4)
plt.tight_layout()
plt.grid(alpha=0.35)
plt.annotate('Source: ELCON A/S',
             xy=(1.0, -0.25),
             color='grey',
             xycoords='axes fraction',
             ha='right',
             va="center",
             fontsize=10)
plt.show()
plt.savefig("./Results/Figures/1_0_costs.png")
plt.savefig("./Results/Presentation/1_0_costs.svg")

# Plot distribution of budget_revenue, sales_estimate_revenue, production_estimate_revenue and final_estimate_revenue
plt.figure(figsize=(10, 5))
sns.kdeplot(data=dfData, x='budget_revenue', label='budget revenue')
sns.kdeplot(data=dfData, x='sales_estimate_revenue', label='sales estimate revenue')
sns.kdeplot(data=dfData, x='production_estimate_revenue', label='production estimate revenue')
sns.kdeplot(data=dfData, x='final_estimate_revenue', label='final estimate revenue')
# limit x-axis to cover 99.99% of the data
plt.xlim(-10, dfData['budget_revenue'].quantile(0.9999999999))
plt.xlabel("Revenue (mDKK)")
plt.ylabel("Density")
# legend below x-axis label
plt.legend(loc='upper center', bbox_to_anchor=(0.5, -0.1), ncol=4)
plt.tight_layout()
plt.grid(alpha=0.35)
plt.annotate('Source: ELCON A/S',
             xy=(1.0, -0.25),
             color='grey',
             xycoords='axes fraction',
             ha='right',
             va="center",
             fontsize=10)
plt.show()
plt.savefig("./Results/Figures/1_1_revenue.png")
plt.savefig("./Results/Presentation/1_1_revenue.svg")


# Calculate risks and other variables
def calculate_risk(group):
    if group['contribution_scurve_diff'].isna().any() or group['contribution_cumsum'].isna().any():
        group['risk'] = np.nan
    else:
        X = group[
            ['revenue_scurve_diff', 'costs_scurve_diff', 'billable_rate_dep']]
        y = group['contribution_scurve_diff']
        model = LinearRegression().fit(X, y)
        residuals = y - model.predict(X)
        group['risk'] = residuals*group['budget_costs']
    return group


dfData = dfData.groupby('job_no', group_keys=False).apply(calculate_risk)

# Show jobs with highest risk
dfData.sort_values('risk', ascending=False).head(15)

# Plot sum of risk by date for each department
plt.figure(figsize=(10, 5))
sns.lineplot(x='date', y='risk', hue='department', data=dfData)
plt.xlabel("Date")
plt.ylabel("Risk")
# legend below x-axis label
plt.legend(loc='upper center', bbox_to_anchor=(0.5, -0.1), ncol=3)
plt.tight_layout()
plt.grid(alpha=0.35)
plt.annotate('Source: ELCON A/S',
             xy=(1.0, -0.15),
             color='grey',
             xycoords='axes fraction',
             ha='right',
             va="center",
             fontsize=10)
plt.show()
plt.savefig("./Results/Figures/1_2_risk.png")
plt.savefig("./Results/Presentation/1_2_risk.svg")

# Select random job and plot risk
job_no = dfData['job_no'].drop_duplicates().sample().values[0]
dfData[dfData['job_no'] == job_no].plot(x='date', y='risk', figsize=(10, 5))
plt.xlabel("Date")
plt.ylabel("Risk")
plt.tight_layout()
plt.grid(alpha=0.35)
plt.annotate('Source: ELCON A/S',
             xy=(1.0, -0.15),
             color='grey',
             xycoords='axes fraction',
             ha='right',
             va="center",
             fontsize=10)
plt.show()

# Calculate total costs at the end of the job
dfData['total_costs'] = dfData.groupby('job_no')['costs_cumsum'].transform('last')
dfData['total_contribution'] = dfData.groupby('job_no')['contribution_cumsum'].transform('last')
dfData['total_margin'] = dfData['total_contribution'] / dfData['total_costs']

# Calculate contribution margin as contribution_cumsum / costs_cumsum
dfData['contribution_margin'] = dfData['contribution_cumsum'] / dfData['costs_cumsum']

# Calculate share of labor cost, material cost and other cost cumsum
dfData['labor_cost_share'] = dfData['costs_of_labor_cumsum'] / dfData['costs_cumsum']
dfData['material_cost_share'] = dfData['costs_of_materials_cumsum'] / dfData['costs_cumsum']
dfData['other_cost_share'] = dfData['other_costs_cumsum'] / dfData['costs_cumsum']

# Omit labor_cost_cumsum, material_cost_cumsum and other_cost_cumsum
dfData.drop(columns=['costs_of_labor_cumsum', 'costs_of_materials_cumsum', 'other_costs_cumsum'], inplace=True)

# Function to set to NA if NaN, inf or -inf
def set_na(x):
    if np.isnan(x) or np.isinf(x) or x == -np.inf:
        return np.nan
    else:
        return x


# Set total_margin, contribution_margin and progress to NA if NaN, inf or -inf
dfData['total_margin'] = dfData['total_margin'].apply(set_na)
dfData['contribution_margin'] = dfData['contribution_margin'].apply(set_na)
dfData['progress'] = dfData['progress'].apply(set_na)

### Encode categorical variables ###
dfData['wip'] = (dfData['status'] == 'wip').astype(int)
dfData['dep_505'] = (dfData['department'] == '505').astype(int)
dfData['posting_group_projekt'] = (dfData['job_posting_group'] == 'PROJEKT').astype(int)

# Convert to categorical
for col in ['responsible', 'address', 'cvr', 'customer', 'job_no']:
    dfData[col] = dfData[col].astype('category')

### Text Processing ###
# Step 1: Filter out the latest description for each job_no
dfDesc = dfData.sort_values('date').groupby('job_no').last().reset_index()
dfDesc = dfDesc[['job_no', 'description']]
dfDesc = dfDesc[dfDesc['description'] != ""]

# Step 2: Preprocess text
stemmer = DanishStemmer()
stop_words = stopwords.words('danish')


def preprocess(text):
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)
    text = re.sub(r'\d+', '', text)
    text = ' '.join([stemmer.stem(word) for word in text.split() if word not in stop_words])
    return text.strip()


dfDesc['description'] = dfDesc['description'].apply(preprocess)

# Step 3 and 4: Convert to Document-Term Matrix and remove sparse terms
vectorizer = CountVectorizer(min_df=0.01, max_df=0.15)
X = vectorizer.fit_transform(dfDesc['description'])
df_matrix = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())

# Step 5: Find and plot the most frequent terms
term_frequencies = df_matrix.sum(axis=0).sort_values(ascending=False)
plt.figure(figsize=(10, 5))
sns.barplot(x=term_frequencies.index, y=term_frequencies.values)
plt.xticks(rotation=90)
plt.xlabel("Terms")
plt.ylabel("Frequency")
plt.tight_layout()
plt.grid(alpha=0.35)
plt.annotate('Source: ELCON A/S',
             xy=(1.0, -0.35),
             color='grey',
             xycoords='axes fraction',
             ha='right',
             va="center",
             fontsize=10)
plt.show()
plt.savefig("./Results/Figures/1_3_description.png")
plt.savefig("./Results/Presentation/1_3_description.svg")

# Step 6: Append the Document-Term Matrix to the original DataFrame
dfDesc.reset_index(drop=True, inplace=True)
df_matrix.reset_index(drop=True, inplace=True)
processed_data = pd.concat([dfDesc[['job_no']], df_matrix], axis=1)

# Left join with the original DataFrame
dfData = pd.merge(dfData, processed_data, on="job_no", how="left")

# Remove description from dfData
dfData.drop(columns=['description'], inplace=True)

### Missing Data Analysis ###
# Calculate missing values
dfMissing = dfData.isna().sum().reset_index()
dfMissing.columns = ['column', 'missing']

# Calculate missing percentage
dfMissing['missing_pct'] = dfMissing['missing'] / dfData.shape[0]

# Sort by missing percentage
dfMissing.sort_values('missing_pct', ascending=False, inplace=True)

# Plot missing percentage above 0%
plt.figure(figsize=(10, 5))
sns.barplot(x=dfMissing[dfMissing['missing_pct'] > 0]['column'],
            y=dfMissing[dfMissing['missing_pct'] > 0]['missing_pct'])
plt.xticks(rotation=90)
plt.xlabel("Columns")
plt.ylabel("Missing Percentage")
plt.tight_layout()
plt.grid(alpha=0.5)
plt.rcParams['axes.axisbelow'] = True
plt.annotate('Source: ELCON A/S',
             xy=(1.0, -0.9),
             color='grey',
             xycoords='axes fraction',
             ha='right',
             va="center",
             fontsize=10)
plt.show()
plt.savefig("./Results/Figures/1_4_missing.png")
plt.savefig("./Results/Presentation/1_4_missing.svg")

### Split test and train ###
# Sample 80% of the jobs for training
lJobNoTrain = dfData['job_no'].drop_duplicates().sample(frac=0.8)
dfData['train'] = dfData['job_no'].isin(lJobNoTrain).astype(int)

# Save dfData to parquet file
dfData.to_parquet("./dfData.parquet")
